package maidsafe.routing.protobuf;

message layer {
  required int32 identifier = 1;
  required bytes message = 2;
}

message Endpoint {
  required bytes ip = 1;
  required int32 port = 2;
}

message Contact {
  required bytes node_id = 1;
  optional Endpoint endpoint = 2;
  repeated bytes local_ips = 3;
  optional int32 local_port = 4;
  optional Endpoint rendezvous = 5;
  optional bool tcp443 = 6;
  optional bool tcp80 = 7;
  optional bytes public_key_id = 8;
  optional bytes public_key = 9;
  optional bytes other_info = 10;
}

message ConfigFile {
  repeated Contact contact = 1;
  optional bytes private_key = 2;
  optional bytes node_id = 3;
}

// Message wrappers 


message Message {
  required bytes source_id = 1; // in response this is name of data
  required bytes destination_id = 2; // in request (except direct) this is the name of the data
  optional bool  cachable = 3;
  required bytes data = 4; // serialised protobuf - possibly encrypted in upper layers 
  optional bool direct = 5; // goes to a single node !! if proxy is set then to here then we transmit to endpoint 
  required bool response = 6;
  required int32 type = 7; // type of data - used in signal to upper layers with (type, data)
  optional bool failure = 8; 
  optional Endpoint proxy = 9;
}

message ConnectRequest {
  optional bool client = 1;
  required Contact contact= 2;
}

message ConnectResponse {
  required bool answer = 1;
  required Contact contact = 2;
}

message FindNodesRequest {
  optional int32 num_nodes_requested = 1;
}

message FindNodesResponse {
  repeated bytes nodes = 1;
}
